tests/cli/test_datasource_snowflake.py

```diff
@@ -0,0 +1,133 @@
+        "xsmall",
+        "public",
+        "password",
+    ]
+
+    credentials = _collect_snowflake_credentials(None)
+
+    assert credentials == {
+        "drivername": "snowflake",
+        "database": "default_db",
+        "host": "ABCD.us-east-1",
+        "password": "password",
+        "query": {"role": "public", "schema": "default_schema", "warehouse": "xsmall"},
+        "username": "user",
+    }
+
+
+@patch("click.prompt")
+def test_snowflake_sso_credentials(mock_prompt):
+    mock_prompt.side_effect = [
+        "2",
+        "user",
+        "ABCD.us-east-1",
+        "default_db",
+        "default_schema",
+        "xsmall",
+        "public",
+        "externalbrowser",
+    ]
+
+    credentials = _collect_snowflake_credentials(None)
+
+    assert credentials == {
+        "drivername": "snowflake",
+        "database": "default_db",
+        "host": "ABCD.us-east-1",
+        "connect_args": {"authenticator": "externalbrowser",},
+        "query": {"role": "public", "schema": "default_schema", "warehouse": "xsmall"},
+        "username": "user",
+    }
+
+
+@patch("click.prompt")
+def test_snowflake_key_pair_credentials(mock_prompt, basic_sqlalchemy_datasource):
+    database_key_path_pass = file_relative_path(
+        __file__, "../test_fixtures/database_key_test.p8"
+    )
+
+    mock_prompt.side_effect = [
+        "3",
+        "user",
+        "ABCD.us-east-1",
+        "default_db",
+        "default_schema",
+        "xsmall",
+        "public",
+        database_key_path_pass,
+        "test123",
+    ]
+
+    credentials = _collect_snowflake_credentials(None)
+
+    assert credentials == {
+        "drivername": "snowflake",
+        "database": "default_db",
+        "host": "ABCD.us-east-1",
+        "private_key_path": database_key_path_pass,
+        "private_key_passphrase": "test123",
+        "query": {"role": "public", "schema": "default_schema", "warehouse": "xsmall"},
+        "username": "user",
+    }
+
+    # making sure with the correct params the key is read correctly
+    basic_sqlalchemy_datasource._get_sqlalchemy_key_pair_auth_url(
+        "snowflake", deepcopy(credentials)
+    )
+
+    # check that with a bad pass phrase an informative message is returned to the user
+    credentials["private_key_passphrase"] = "bad_pass"
+    with pytest.raises(DatasourceKeyPairAuthBadPassphraseError) as e:
+        basic_sqlalchemy_datasource._get_sqlalchemy_key_pair_auth_url(
+            "snowflake", deepcopy(credentials)
+        )
+
+    assert "passphrase incorrect" in e.value.message
+
+    # check that with no pass the key is read correctly
+    database_key_path_no_pass = file_relative_path(
+        __file__, "../test_fixtures/database_key_test_no_pass.p8"
+    )
+    credentials["private_key_path"] = database_key_path_no_pass
+    credentials["private_key_passphrase"] = ""
+    basic_sqlalchemy_datasource._get_sqlalchemy_key_pair_auth_url(
+        "snowflake", deepcopy(credentials)
+    )
```

Comment:
What are we testing here, and what is the private key material included for the test?

Reply:
Oops! I think I must have forgotten to add assertions, when adding the code I just wanted to make sure it wasn't throwing exceptions but now I added some assertions.

I included 2 private keys for testing (which I generated just for this purpose), one has no password, the other one has test123 as a password, wanted to add some tests to make sure that users  could use both types of keys.